Feature/member

kidocchy opened the pull request:
プログラムからSQLをいじるのに慣れていなくて，検証環境づくり，バグとりに時間がかかりました。すみませぬ。

ユーザー登録，呼び出しができるようになりました。
自宅のSpigot1.18 / SQLiteで動作確認済み。

Profundus.javaのonEnable()でcreateTableを呼び出し。SQLのテーブルを準備。

LoginEvent.javaで，onPlayerJoinEventをlisten
その中で，User.getUser(Player,Boolean)でデータの呼び出し(なければ作成)と，
動作確認用に，LastLoginをメッセージで出力。

Userデータは，とりあえず初回ログイン，最終ログインなどを記録。
コミュニケーション用にmainLanguage, subLanguageのプロフィールを入れる場所は作ってありますが，
入出力の部分はないので，今後の課題。ゆくゆくは他のプロフィール項目と合わせてWebインターフェイスを用意してもいいのかなと思っています。
また，UserデータはSQLに格納しているのはもちろん，
処理高速化のために，プラグイン稼働中はインメモリで保持しています。
なので，ユーザ情報の呼び出しがあれば，インメモリを検索->なければSQLを検索->なければ新規作成
という工程になっています。

今のところデータの削除機能は実装していないので，今後の必要に応じての課題かなと。

DatabaseUtil.java
PFID.java
User.java
はその処理用バックエンド。

Commit:
test commit

Commit:
ちょっと変えてみるテスト

Commit:
for device synchronization

Commit:
for device sync

Commit:
for device sync

Commit:
とりあえずコードかいた。テストとコメントはこれから。

Commit:
ユーザー登録まわり完成

とりあえずユーザー登録，呼び出し関係は動くようになりました。
自宅サーバSpigot1.18/SQLite

Commit:
onEnableとEventListnerに追記

kumitatepazuru committed:
Merge branch 'develop' into feature/member

kumitatepazuru commented:
@kidocchy いい忘れてました。初回なので（ぱずる流）PRのやり方〜
一旦はGithub Actionsがpassedになるまで修正をお願いします。今のところはビルドチェックとJavadocの記入漏れの確認をしてます。もし、Actionsが動かなかったらConfrictしていると思うのでPRのメッセージに沿って修正をお願いします。
Actionsがpassedになったら、実際に自分が簡単に挙動の確認等をします。それで、問題がなさそうでしたらmergeという形をとっています。

現在はActionsがfailedになっているので、Detailsから問題部分を修正してcommitをお願いします。

kidocchy commented:
@kumitatepazuru はいー
すみません調べながらやってみます。

Commit:
修正ついでにグループDB機能追加

Commit:
バグ鳥・JAVADOC

Commit:
JAVADOC書き殴り

Commit:
JAVADOC書き殴り

Commit:
JAVADOC書き殴り

kumitatepazuru commented on src/main/java/tech/inudev/profundus/listener/LoginEvent.java:
```diff
@@ -0,0 +1,35 @@
+		User user = User.getByPlayer(e.getPlayer(), true);
+		String lastLoginStr = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss").format(LocalDateTime.ofInstant(user.getLastLogin(), ZoneId.of("Asia/Tokyo")));
+		user.updateLastLogin();
+		user.getPlayer().sendMessage("Profundus[LOGIN]/DATA_SOURCE:" + user.getSource());
```
プレイヤーにこのような情報は表示させなくても大丈夫です。（下の行のやつも同様）

kumitatepazuru commented on src/main/java/tech/inudev/profundus/listener/LoginEvent.java:
```diff
@@ -0,0 +1,35 @@
+		User user = User.getByPlayer(e.getPlayer(), true);
+		String lastLoginStr = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss").format(LocalDateTime.ofInstant(user.getLastLogin(), ZoneId.of("Asia/Tokyo")));
+		user.updateLastLogin();
+		user.getPlayer().sendMessage("Profundus[LOGIN]/DATA_SOURCE:" + user.getSource());
```
プレイヤーにこのような情報は表示させなくても大丈夫です。（下の行のやつも同様）

kidocchy replied:
・本番環境には，いらないと思っていました。
・開発中の動作確認用ですので，デバッグ用に適切な出力先(loggerがいいんでしょうか？）があれば，そちらへの出力でいいのかなとも思います。

kumitatepazuru replied:
ここでのPRは完成品を提出するためのPRでこのようなデバッグ用コードは削除してPRしてほしいというのが自分の気持ちです。これに関連することがあればその都度またテストコードを書いて、っていうのがいいのかなって思ってます。
リリース時に皆がこのコードの存在を忘れていて”バグ的存在”になっちゃうのも厄介ですし。

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/DatabaseUtil.java:
```diff
@@ -40,6 +42,7 @@ public static void connect() {
             connection = DriverManager.getConnection(databaseUrl, configHandler.getDatabaseUsername(), configHandler.getDatabasePassword());
             connection.setAutoCommit(false);
         } catch (SQLException e) {
+            System.out.println("Couldn't connect DB");
```
Plugin.getLogger()を使ってください。（Profundus.getLogger()で行けたはず）
https://jd.papermc.io/paper/1.18/org/bukkit/plugin/Plugin.html#getLogger()
ほかの`System.out.println`も同様の対処をお願いします。

kidocchy replied:
getLoggerまで辿り着けずこうなっていました。Profundus.getLogger()試してみます，

kidocchy replied:
Profundus(ext.JavaPlugin).getLogger()はstaticから呼ぶと怒られたので，
Profundusクラスで，public static Logger loggerというクラス変数を用意して，onEnable内でgetLogger()で格納しています。
ですので，Profundus.logger.log(Level,String)
で呼べます。

kumitatepazuru replied:
@kidocchy commit後で申し訳ないのですがもともと`getInstance().getLogger()`で行けるのでそちらを使っていただきたいです。申し訳ない...!

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/DatabaseUtil.java:
```diff
@@ -207,4 +211,533 @@ private static void commitTransaction() {
+ * @author kidocchy
+ *
+ */
+    public enum Table{
```
今後これは増えていくことが予想されるのでenumを`USER("USER")`のような形にして`valueOf(String)`で参照できるようにしてください。
なお、その時下の`stringToEnum`はvalueOfで代替できるので削除してください。

kidocchy replied:
Stringでいいところを値の選択肢を制限する目的でenumを使ってみたのですが，USER("USER")の括弧内がなくても，
USER,
PFGROUP,
みたいに列挙するだけで，Stringとのやりとりをスマートに記述できる方法ってないもんですかね。(@Getter)みたいな。
なければ，ご指摘の通り試してみます

kumitatepazuru replied:
そもそも`valueOf`はかっこつけてなくても行けるみたいですね
https://techacademy.jp/magazine/33623

kidocchy replied:
<img width="798" alt="Screen Shot 2022-07-01 at 19 49 35" src="https://user-images.githubusercontent.com/78601427/176880732-0db4eb9d-ebca-4905-bba5-683137c72a40.png">
これでいきます。

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/DatabaseUtil.java:
```diff
@@ -207,4 +211,533 @@ private static void commitTransaction() {
+     * @param dropIfExists true=>存在しても再作成(初期化) 
+     * @return success?
+     */
+    public static Boolean createTable(Table tableName, Boolean dropIfExists) {
```
上の`createMoneyTable`も合体してほしいです。
存在しても再作成は手動で行う予定ですのでその仕様は削除してください

kidocchy replied:
合体了解

再作成は，もう少しデータベースのテーブルの仕様が固まるまで残しておきたいのですがどうでしょう。
当然，本番環境が動き出せば不要，というか，使えなくするべき機能だと思っています。

kumitatepazuru replied:
了解です。では、コメントで`TODO: `から始まる分としてリリース時に削除する事の記入をお願いします。
また、Profundusに書かれているこの関数を呼び出す部分のこのフラグをfalseにしてください。

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/DatabaseUtil.java:
```diff
@@ -207,4 +211,533 @@ private static void commitTransaction() {
+	    }
+    }
+/*
+    static Boolean insertItemEntry(PFItem item, UUID newPFID, String note) {
```
テスト用のコメントでしたら削除をお願いします。

kidocchy replied:
これは，別のブランチにすべきところを使い方がわからず，作り始めたアイテム管理の実装の破片です。
とりあえず消します。

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/DatabaseUtil.java:
```diff
@@ -207,4 +211,533 @@ private static void commitTransaction() {
+ * 確実にconnectedなインスタンスを取得する。
+ * @return
+ */
+    static Connection getConnected(){
```
getterになるように`getConnection`に改名をお願いします。

kidocchy replied:
この部分は，DB接続が切れていた時に再接続するためのコードです。

で，ここは相談ですが，元々の実装では，定期的にPINGを送ってコネクションを維持しているようです。
それが確実に機能するのであれば，この接続チェックとか噛まずに，インスタンスから直接Connectionをとれば済む話なのですが・・・。

１．この再接続コードは必要か，不要か
２．ぐぐると，コネクションプールを使うのが一般的と出てくるが，実装の予定はあるか

kumitatepazuru replied:
再接続コードに関して、なくてもpingを定期的に打つことによって切断されないようにしているため問題はないと踏んでいますが、万が一のため（SQLのメンテ等とかがあったりするからその時の再接続とか）そのままでお願いします。

コネクションプールはもともとclassがstaticになっていてpluginがenableになるときに接続、disableになるときに切断という処理になっているはずなのですでに実装されているかと。（やり方が違っていたら教えていただきたいです）

kidocchy replied:
コネクションプールに関して，確かに調べてみると，現実装でも，一度接続したものを使い回すという意味では一種のコネクションプーリングと言うようですね。

私のイメージとしては，プールしておく，つまり溜めておく場所に，あらかじめ接続されたインスタンス的なものを５個とか１０個とか持っておいて，それを管理するクラスがある。そこにgetConnectionみたいに要求すると上限数までそのうち１つを貸し出してくれて，クエリを実行して，closeすると返却される。
背景には，
・タスクごとにconnect/closeするのが本当はいいらしい
・実際問題，そんなことしていたらオーバーヘッドで遅すぎる
というのがあるみたいです。
プールしておくことで，
・複数接続・接続数上限の管理
・DBサーバーメンテ時など，待たせておくことが可能
らしいデス。

マイクラのサーバーがシングルスレッドだと聞いたことはあるのですが，一つのconnectionを使い回して，処理が増えてきた時にどこまで確実に処理できるのかな？というのが疑問なところです。

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/DatabaseUtil.java:
```diff
@@ -207,4 +211,533 @@ private static void commitTransaction() {
+     * @param query
+     * @return success?
+     */
+    static ResultSet select(Table tableName, String query){
```
ほかにも使用しているselectがあるのでそれもこの関数へ置き換えをお願いします。

kidocchy replied:
>ほかにも使用しているselectがある
の意味がピンとこないのですが，
どの部分でしょうか。

kumitatepazuru replied:
https://github.com/inudevtech/Profundus/pull/79/files/44de3451f4326c3a8f8313ef9397b1e001430454#diff-e01330d123e0e0d8448b398d0aadcd4c5a9aa4a294deb2cceb95dc5424bc386fL137
ここですかね（もしかしたらほかにもあるかも）

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/PFGroup.java:
```diff
@@ -0,0 +1,145 @@
+ * @author kidocchy
+ *
+ */
+public class PFGroup extends PFAgent{
```
グループ機能はLuckPermsのAPIを使用する予定ですのでこちらはいったんなしでお願いします。

kidocchy replied:
この実装によって，実現しているorしようとしている機能として，
User(個人)とGroup(法人)をPFAgentから継承することで，
1. 区画やアイテムの所有者として個人法人どちらも入れられるようになっている
(口座も法人で持てるようにする構想もどこかで見かけたような・・・)

class PFChunk
   PFAgent owner = (User or Group) hogehoge;
-----------
  owner.sendMessage(....) みたいなことが既に可能。

2. SQLサーバーという，マイクラ本体とは別でデータを管理している=SQLサーバ側の組み方でデータ保全性の確保をしやすい
の部分が大きいかなと思っています。

現状そのAPIの方が機能は豊富でしょうし，それに乗っかることは異論ないですが，
API側の都合に振り回されないよう，
少なくとも大元のデータはこちらのデータベースで管理した方がいいように思います。

ということで，どっちかがラッパーとして働くなどメンテナンスはちょい大変だと思いますが，
独自のグループクラスを持っておくことは開発の青写真としても，今後の展望を考えても有用と思いますが，
いかがでしょう。

kumitatepazuru replied:
1に関しては気持ちはわかるのですがLPのAPIが強力すぎて運営、という観点でも非常に便利なので向こうのAPIから継承という形で実装したほうがいいかもしれないというのが自分の意見です。
2に関してはLPもMySQLを使用したデータ管理が可能で、データ同期などの機能も豊富です。
APIの仕様が変わる、というのはまずありえないと考えてよいと思うので（考えると多分きりがない）LPの実装で行きたいというのが思いです。

正直、このプロジェクトがどのくらいの大きさになるのかが全然わからないので既存システムのメンテをできるだけ減らして新規仕様の作成等に回したい、というのがあります。

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/PFID.java:
```diff
@@ -0,0 +1,93 @@
+ * @author kidocchy
+ *
+ */
+public abstract class PFID{
```
NamespacedKeyに似た実装な気がするのでNamespacedKeyを継承しての実装をお願いしたいです。

kidocchy replied:
ほい。それは知らないやつなので，調べてきます。

kidocchy replied:
調べてきました。
final class NamespacedKeyはプラグイン側ではなくマイクラspigot側にデータを持たせる必要があるときのkeyを生成するものですね。

今回のPFIDの実装は，例えば
https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/persistence/PersistentDataContainer.html
でいうところの

`PersistentDataContainer. <T,Z> void set(NamespacedKey key, PersistentDataType<T,Z> type, Z value)
`
の，Z valueの方を生成・管理する部分になると思いますので，NamespacedKeyとは目的が違います。

細かくは，この先やっていく土地やアイテムの実装で検証しますが，おそらく上記setしておいたデータ(例えば管理ID)を
get( =profundus:pfid=, long)　注： ==内はNamespacedKeyで生成
で取り出して，
今回のPFIDクラスのメソッドgetByPFIDでProfundusプラグインの何らかのインスタンスを取得，
そこから処理をコールしていく
ということになるのでしょう。
今後使いそうなので，PFIDを入れ込む用のNamespacedKeyはクラス内で作っておくことにしました。

PS.別スレッドで質問したITEMデータの件ですが，まだ返信追えてないのですがこれで解決しそうですね。

kumitatepazuru reviewed (changes requested):
修正ありがとうございます。数点変更していただきたい部分があるので申し訳ないですが、さらに修正をお願いします。

kidocchy commented:
@kumitatepazuru 
先に聞けばよかったのですが。
皆さんテスト用コードってどうしてますか？

指摘してもらった中に、
本番環境では確かにいらないんだけど、
今はまだ仕様が固まってない中、
アルファ版になる前ぐらいまで残しておきたいものがちょいちょいあるのですが。

kumitatepazuru commented:
今のところ特に決まってはいません。
一応、皆さんはPR前にJUnitを使ったテストコードは書いていますが、PR時までには消してます。
そこらへんは盲点だったので後ほど決めようかなと思います。（JUnitなどとプラグインがあまり相性が良くなくて決められてないのも事実）

kumitatepazuru commented:
指摘したもので一旦は残しておきたいものはここにリストアップしてくれれば何かしらの残す等の対策は取りますので連絡をいただければ。

Commit:
指摘事項修正

Commit:
JAVADOC

Commit:
s/\.logger\(\)\/.//.getInstance\(\)\.getLogger\(\)\./

kumitatepazuru commented on src/main/java/tech/inudev/profundus/utils/User.java:
```diff
@@ -0,0 +1,240 @@
+ * @author kidocchy
+ *
+ */
+public class User extends PFAgent{
```
このclass、BukkitのPlayerを継承して作ったほうが後々何かと楽そうなのでそっちにしてもいい気がするんですけどどうでしょうか

kidocchy replied:
Playerはinterfaceのようなので、
Extendではなくimplementになりますよね。
であれば、
Playerのメソッドの挙動を変えたい時に、
Overrideで変えたり、
User.foo()
User.getPlayer().bar()
という呼び出しの使い分けをしなくていいというのがメリットになると思いますが、認識あってます？

別の問題として、Userクラスは、プレイヤーがログインしていなくても、売買などでインスタンスを使えるように考えています。
その際、Bukkit.player関係は呼び出せないはずなので、その処理（NullPointerとかになるのかな？）はなんとかしなきゃな、と思っていたところです。その実装は、interfaceをimplementしない方が楽な気もしますね、、、。

tererun replied:
私の個人的な理由は`User#getPlayer`とか`User#getOfflinePlayer`とかそういうのを実装すべきかなあと思ったりしています。
kidocchyさんがいうようにOfflineになった際の処理が面倒そうということから上記な形にすべきかなあと思ったりしてます。
あと挙動忘れたんですが（違ったらすみません）プレイヤーがリログしたらPlayerインスタンスが違うものになっていた気がしなくもないので、フィールド変数にPlayerを入れるのではなく毎回`Bukkit.getPlayer`をしてあげた方がいい気もします。

kidocchy replied:
@kumitatepazuru 目的に立ち返って考えると，Playerはマイクラ内での挙動, Userはユーザー管理というところになると思うので，確かに被っているところもありつつ，とりあえず別でスタートしても良いように思いました。
上でも書きましたが，Userから呼べるメソッドは，対象者がログイン中じゃなくてもいいとしておく，というのが大きいところでしょうかね。

kumitatepazuru replied:
一週間ほど前に送ったはずだったのですがメッセージが送信されていなかったみたいなので再度送信します。遅れてしまい申し訳ないです。
そうでした、実装はNMSのCraftPlayerでしたね。
継承してほしい理由としては、似てるところがものすごい多いというのがあって、オフラインのとき等の問題はCraftOfflinePlayerを継承してやるってことでもいいのかなと思ってます。
ユーザー管理という点ではPlayer(CraftPlayer)もそういう立ち位置にあるものなのでほぼ一緒なのかな、と。

kidocchy replied:
先週区画管理のあたりのコードを書いていたのですが，このUserクラスの役割はやはりProfundusとしてのユーザー管理＝SQLとのやりとりがメインなのではと思いました。ですので，区画の所有者として，このクラスのインスタンス(に対応したID)が格納できれば，私としては中身の実装は特にこだわりはありません。
tererun案→ getPlayer() / getOfflinePlayer() を実装してもいいですし，
第２案→ getPlayer()の中で判定して，CraftPlayer / CraftOfflinePlayerを適宜返すでもいいですし，
pazuru案→ 全部implementしてもいいですし。メンテ大変そうだけど。
要は，使う人がどう使いたいか，ですよね。
今のところ想定しているのは，

１。区画所有をセットするときに，例えばイベント駆動であれば
PFChunk#setOwner( User.getByPlayer(e.getPlayer() ) -> void

２。区画所有を判定するときに
User#isOwner( PFChunk.getByChunk( Player.getChunk() ) ) -> boolean

みたいに使うのかなぁと思っています。
ということで，どう使いたいか，という例をもらえた方が実装を考えやすいかなと思います。

kumitatepazuru replied:
@kidocchy じゃあ作業量とかを考えて一旦てれるんさんのget~()を実装で行きましょうか。
それでやってってプログラムが読みづらくなってきた！とかになったら適宜変更するって感じで。

kumitatepazuru reviewed (changes requested):
遅れました。度々すみません、一つ提案があるので確認していただければと思います。

tererun commented on src/main/java/tech/inudev/profundus/utils/PFAgent.java:
```diff
@@ -0,0 +1,36 @@
+package tech.inudev.profundus.utils;
```
パッケージおじさんです、こんにちは。
utilsパッケージは基本的にインスタンスを作らないクラスをまとめるという形が正規の形になります。
@kumitatepazuru 君の考え的にはあまりパッケージを増やしたくないという考えだったので「これはこのパッケージを作って...ごにょごにょ」みたいなことができないのでなんとも言えませんが、私的にはUser系の処理なので「users」とかにまとめたい気持ちです。

kidocchy replied:
パッケージの管理どうしているのかわからんので，とりあえずutilに入れていました。
ゆくゆくはデーターベースいじる系（ユーザー・お金・所有権）はそれでまとめたらいいのかなと思っていたところでした，

kumitatepazuru replied:
この際なのでデータベースいじる系のパッケージ作っちゃいましょうか
utilがごちゃごちゃしてもあれですし。
名前はおまかせしますｗ

kidocchy replied:
パッケージ変更しようと思っていたのですが，utilsだけutilではなく複数形なのなんか文化とか意味とかあります？

あと，...profundus.util(s).databaseと
...profundus.databaseと
どちらが好みですか？

kumitatepazuru replied:
なんでutilsなのかは自分も記憶にないですがおそらく単数でOKだと思います（間違ったことは言えないのでパッケージおじさん召喚しときます @tererun )
パッケージに関しては自分は下のほうが嬉しいです（長いのはちょっとやだ）

tererun replied:
呼ばれたパッケージおじさんです
正直どちらでもいいですが、私はいつも可算名詞かどうかで決めるようにしています。
パッケージの中には複数のクラスやインターフェースが入る分けですから、複数形、という気持ちです。

kumitatepazuru replied:
じゃあそのままで行きましょ〜（圧倒的適当感）

tererun commented on src/main/java/tech/inudev/profundus/utils/User.java:
```diff
@@ -0,0 +1,240 @@
+package tech.inudev.profundus.utils;
```
これも同様にutilsっぽい扱いじゃないので〜って感じです。

tererun commented:
色々気になった点を書きました　コメントApproveするのわすれてました

Commit:
修正。utils複数形問題は未対応

・databaseパッケージを作成，データベースいじる系は移動。
・getPlayer() / getOfflinePlayer()実装

Commit:
JAVADOC

kumitatepazuru reviewed (changes requested):
developとのconflictの修正をお願いします

Commit:
fix

Commit:
Merge branch 'develop' into feature/member

Commit:
fix

Commit:
fix3

Commit:
goodsテーブルの作成メソッドを統合

ついでにJAVADOC

kidocchy commented:
gitむずかしい・・・
eclipseむずかしい・・・

Commit:
リファクタリング。

Database系が太ってきたので，

１。DatabaseUtilには接続関係と，共通で使う構文

２。それぞれのtableをいじるためのSQLは
「DBUホニャララ」

にクラス分割

Commit:
JAVADOC...

kumitatepazuru reviewed (changes requested):
修正ありがとうございます。

UserクラスのLanguageとLastLoginに関してなのですが、何度も申し訳ないですがPlayerクラスの[getLocale](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/Player.html#getLocale())とOfflinePlayerクラスの[getLastPlayed](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/OfflinePlayer.html#getLastPlayed())で代用できると思うので削除をお願いしたいです。

kidocchy commented:
反応が悪くすみません。
languageは，例えば私はenでマイクラしてますが，コミュニケーションはjpでOKです。
そういった，プロフィール情報のための準備です。

lastloginは，webインターフェイスとかからログインした時間も記録するための準備です。
メタバースなので，ゲームで全部やれよ，という気もしますが，ゲームプレイ＋webインターフェイスでフリマ，というのもありなのかな，と構想してます。

という意図を伝えたところで，要らないということであれば削除しますがどうでっしゃろ？

kumitatepazuru commented:
じゃあ残しましょうか（）
ただ、SubLanguageに関しては実態使わないと神に誓ってでも言えるのでこれだけ削除をお願いしたいです
お手数おかけします

kumitatepazuru committed:
Delete docs directory

Commit:
第二言語プロフィール部分削除

kumitatepazuru reviewed (approved):
ありがとうございます。問題なさそうなのでmergeします。
何度も修正をしてくださりありがとうございました!
